tests/test_task.py

```diff
@@ -325,6 +319,53 @@ def testTimeMethod(self):
             new_meta = yaml.safe_load(y)
         self.assertEqual(new_meta, addMultTask.metadata)
 
+    def test_annotate_failure_generic_exception(self):
+        """Test annotating failures in the task metadata when a non-Task
+        exception is raised.
+        """
+        task = AddMultTask()
+        msg = "something failed!"
+        error = ValueError(msg)
+        with self.assertLogs("addMult", level="WARNING") as cm:
+            task._annotate_failure(error)
+        self.assertIn(msg, "\n".join(cm.output))
+        self.assertEqual(task.metadata["failure_message"], msg)
+        self.assertEqual(task.metadata["failure_type"], "ValueError")
+        self.assertNotIn("failure_metadata", task.metadata)
+
+    def test_annotate_failure_task_exception(self):
+        """Test annotating failures in the task metadata when a Task-specific
+        exception is raised.
+        """
+        class TestError(pipeBase.RepeatableQuantumError):
+            @property
+            def metadata(self):
+                return {"something": 12345}
+
+        task = AddMultTask()
+        msg = "something failed!"
+        error = TestError(msg)
+        with self.assertLogs("addMult", level="WARNING") as cm:
+            task._annotate_failure(error)
+        self.assertIn(msg, "\n".join(cm.output))
+        self.assertEqual(task.metadata["failure_message"], msg)
+        result = "test_task.TaskTestCase.test_annotate_failure_task_exception.<locals>.TestError"
+        self.assertEqual(task.metadata["failure_type"], result)
+        # TODO: can we make this work without it being in "`scalars`?
```

Comment:
TaskMetadata is trying to be a bit like PropertyList but in Python. The internal pydantic model separates list items from scalar items but you aren't really meant to be looking inside that (but pydantic doesn't let me call them private properties).

Reply:
Why isn't the test doing:

```python
self.assertEqual(task.metadata["failure_metadata"]["something"], 12345)
```
?

Reply:
Also, if you want to write TaskMetadata to FITS headers then copying to PropertyList should work fine. I'm not sure what the use case is.

Reply:
@timj : I didn't know what the API for nested metadata was here; that makes the test cleaner, thank you.

I'd suggest that nested metadata like this not print like this:

```
ipdb> task.metadata["failure_metadata"]
TaskMetadata(scalars={'something': 12345}, arrays={}, metadata={})
```

since that doesn't apply to me that you can just access the `scalars` by name.

Reply:
That's the default stringification of the Pydantic model (effectively `repr`). I think a `__str__` implementation that returned a dict like thing should be fine. TaskMetadata is designed to match PropertyList without the C++.

Reply:
I just did some experimenting and we do have an important asymmetry between `TaskMetadata` and `PropertyList` that we may need to work around (`PropertySet` behaves like `TaskMetadata` and unlike `PropertyList`):
```
t = TaskMetadata()
t["a"] = {"b": 3}
assert isinstance(t["a"], TaskMetadata)
assert t["a"]["b"] == 3
```
but with
```
p = PropertyList()
p["a"] = {"b": 3}
```
accessing `p["a"]` raises `KeyError`, because the assignment flattens the dict out into `p["a.b"] = 3`.  @timj , do you know if there is a dedicated method to reconstruct those flattened mappings that works on all of `PropertySet` / `PropertyList` / `TaskMetadata`?  If not, I think we should add one.

As an aside, while testing this I also discovered that `TaskMetadata(b=3)` just silently swallows the `b=3` and does nothing with it, probably because of something Pydantic-related.

Reply:
PropertyList is not hierarchical but PropertySet and TaskMetadata are. PropertyList probably should not inherit from PropertySet. TaskMetadata is trying to do it all in Python but was never designed for PropertyList emulation, partly because Task metadata was historically a PropertySet.  PropertyList in afw.Exposure has always been an interesting situation since it's sort of trying to emulate a FITS header without saying it emulates a FITS header (and the array values are problematic in FITS and best avoided in general). Ideally we would read back HIERARCH FITS headers into PropertyList and get a real hierarchy as in PropertySet.

Re the question on flattened mappings. You mean you want PropertyList to work like PropertySet and TaskMetadata so that `p["a"]` will return a PropertyList?  

Yes, the TaskMetadata constructor doesn't really want you to pass any arguments into it because the pydantic model gets in the way. It's why the `from_dict` class method exists. `from_metadata` is trying to be the way you convert from a PropertySet to a TaskMetadata.

Reply:
> PropertyList probably should not inherit from PropertySet.

:100: and I once tried to fix this but it was a nightmare.

> You mean you want PropertyList to work like PropertySet and TaskMetadata so that p["a"] will return a PropertyList?

I want a method that does that on all three classes, but I don't need it to be `__getitem__`, since I'm sure making `__getitem__` do that on `PropertyList` would break lots of stuff.  In the other two classes it could just delegate to `__getitem__`.

Reply:
I'd suggest moving this conversation over to [DM-42928](https://jira.lsstcorp.org/browse/DM-42928), once that's been pushed? I think it's more appropriate for that one (and then this ticket can just use whatever is implemented there).